src/Note.cpp

```diff
@@ -0,0 +1,378 @@
+/*************************************************************************/
+/*  Xania (M)ulti(U)ser(D)ungeon server source code                      */
+/*  (C) 1995-2000 Xania Development Team                                 */
+/*  See the header to file: merc.h for original code copyrights          */
+/*                                                                       */
+/*  Note.cpp: notes system                                               */
+/*                                                                       */
+/*************************************************************************/
+
+#include "Note.hpp"
+#include "DescriptorList.hpp"
+#include "TimeInfoData.hpp"
+#include "db.h"
+#include "merc.h"
+#include "string_utils.hpp"
+
+#include <fmt/format.h>
+#include <range/v3/algorithm/find_if.hpp>
+#include <range/v3/view/filter.hpp>
+
+#include <cctype>
+#include <cstdio>
+#include <functional>
+
+static const char *NOTE_FILE = "notes.txt";
+
+bool Note::is_to(const Char &ch) const {
+    if (matches(ch.name, sender_)) {
+        return true;
+    }
+    if (is_name("all", to_list_)) {
+        return true;
+    }
+    if (ch.is_immortal() && is_name("immortal", to_list_)) {
+        return true;
+    }
+    if (is_name(ch.name, to_list_)) {
+        return true;
+    }
+    return false;
+}
+
+bool Note::sent_by(const Char &ch) const { return matches(ch.name, sender_); }
+
+void Note::save(FILE *file) const {
+    fmt::print(file, "Sender {}~\n", sender_);
+    fmt::print(file, "Date {}~\n", date_);
+    fmt::print(file, "Stamp {}\n", Clock::to_time_t(date_stamp_));
+    fmt::print(file, "To {}~\n", to_list_);
+    fmt::print(file, "Subject {}~\n", subject_);
+    fmt::print(file, "Text\n{}~\n", text_);
+}
+
+void Notes::erase(const Note &note) {
+    auto it = ranges::find_if(notes_, [&](const Note &data) { return &data == &note; });
+    if (it != notes_.end())
+        notes_.erase(it);
+}
+Note &Notes::add(Note note) { return notes_.emplace_back(std::move(note)); }
+
+int Notes::num_unread(const Char &ch) const {
+    int notes = 0;
+    for (auto &note : notes_) {
+        if (note.is_to(ch) && !note.sent_by(ch) && note.date_stamp() > ch.last_note) {
```

Comment:
Q: it doesn't matter, but might there be a `ranges` way to iterate & filter the notes? (Just thinking in terms of how it can be done with Java streams).

Reply:
Definitely! Good point, then I can use `count`. Will fix